File: submissions/2121-find-if-path-exists-in-graph/solution.ts
function validPath(n, edges, source, destination) {
    const graph = new Map();

    for (let i = 0; i < n; i++) {
        graph.set(i, []);
    }

    for (const [x, y] of edges) {
        graph.get(x).push(y);
        graph.get(y).push(x);
    }

    let visited = new Set();

    function dfs(node) {
        if (node === destination) {
            valid = true;
            return;
        }

        visited.add(node);

        for (const neighbor of graph.get(node)) {
            if (!visited.has(neighbor)) {
                dfs(neighbor);
            }
        }
    }

    let valid = false;
    dfs(source);

    return valid;
}

File: submissions/2121-find-if-path-exists-in-graph/solution.py
from collections import defaultdict
class Solution:
    def validPath(self, n: int, edges: List[List[int]], source: int, destination: int) -> bool:
        graph = defaultdict(list) 
        seen = {source} 
        
        for [x, y] in edges:
            graph[x].append(y)
            graph[y].append(x)
            
        def dfs(node):
            if node == destination: 
                return True
            
            for neighbour in graph[node]:
                if neighbour not in seen: 
                    seen.add(neighbour)
                    if dfs(neighbour):
                        return True
            return False
        
        return dfs(source)
                
            
